server/executor/queue.go

```diff
@@ -0,0 +1,474 @@
+		TransactionRun: transaction.TransactionRun{ID: job.TransactionRun.ID},
+
+		PollingProfile: pollingprofile.PollingProfile{ID: job.PollingProfile.ID},
+		DataStore:      datastore.DataStore{ID: job.DataStore.ID},
+	}
+
+	q.driver.Enqueue(newJob)
+}
+
+func (q Queue) Listen(job Job) {
+	// this is called when a new job is put in the queue and we need to process it
+	propagator := propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})
+	ctx := propagator.Extract(context.Background(), propagation.MapCarrier(*job.Headers))
+
+	ctx, cancelCtx := context.WithCancel(ctx)
+	q.listenForStopRequests(context.Background(), cancelCtx, job)
+
+	newJob := Job{
+		Headers: job.Headers,
+	}
+	newJob.Test = q.resolveTest(ctx, job)
+	// todo: revert when using actual queues
+	// newJob.Run = q.resolveTestRun(ctx, job)
+	newJob.Run = job.Run
+
+	newJob.Transaction = q.resolveTransaction(ctx, job)
+	newJob.TransactionRun = q.resolveTransactionRun(ctx, job)
+
+	newJob.PollingProfile = q.resolvePollingProfile(ctx, job)
+	newJob.DataStore = q.resolveDataStore(ctx, job)
+
+	// Process the item with cancellation monitoring
+	select {
+	default:
+	case <-ctx.Done():
+		return
+	}
+	q.itemProcessor.ProcessItem(ctx, newJob)
+}
+
+type StopRequest struct {
+	TestID id.ID
+	RunID  int
+}
+
+func (sr StopRequest) ResourceID() string {
+	runID := (test.Run{ID: sr.RunID, TestID: sr.TestID}).ResourceID()
+	return runID + "/stop"
+}
+
+func (q Queue) listenForStopRequests(ctx context.Context, cancelCtx context.CancelFunc, job Job) {
+	if q.subscriptor == nil {
+		return
+	}
+
+	sfn := subscription.NewSubscriberFunction(func(m subscription.Message) error {
+		cancelCtx()
+		stopRequest, ok := m.Content.(StopRequest)
+		if !ok {
+			return nil
+		}
+
+		run, err := q.runs.GetRun(ctx, stopRequest.TestID, stopRequest.RunID)
+		if err != nil {
+			return fmt.Errorf("failed to get run %d for test %s: %w", stopRequest.RunID, stopRequest.TestID, err)
+		}
+
+		if run.State == test.RunStateStopped {
+			return nil
+		}
+
+		return q.cancelRunHandlerFn(ctx, run)
+
+	})
+
+	q.subscriptor.Subscribe((StopRequest{job.Test.ID, job.Run.ID}).ResourceID(), sfn)
+}
+
+func (q Queue) resolveTransaction(ctx context.Context, job Job) transaction.Transaction {
+	if q.transactions == nil {
+		return transaction.Transaction{}
+	}
+
+	tran, err := q.transactions.GetAugmented(ctx, job.Transaction.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return transaction.Transaction{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return tran
+}
+func (q Queue) resolveTransactionRun(ctx context.Context, job Job) transaction.TransactionRun {
+	if q.transactionRuns == nil {
+		return transaction.TransactionRun{}
+	}
+
+	tranRun, err := q.transactionRuns.GetTransactionRun(ctx, job.Transaction.ID, job.TransactionRun.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return transaction.TransactionRun{}
+	}
+	if err != nil {
+		panic(err)
```

Comment:
Philosophical question: if an error happens with a Job, should we just `panic`, bringing down the engine or just drop the `job` and add it to a `dead letter` queue/storage?

Reply:
the only reason for this to happen is because we don't have access to the DB for some reason. if we lost access to DB, or it's locked, or queries are wrong, it feels reasonable to have a panic and restart the service.  I'm not married to this idea though.

While I think it might be a good idea to have a death queue, it might add complexity that we don't neccesariliy need at the moment IMO